database/database.go

```diff
@@ -0,0 +1,151 @@
+package database
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+
+	"github.com/HouzuoGuo/tiedot/db"
+	"github.com/fatih/structs"
+	"github.com/josephroberts/edge-node-manager/config"
+	"github.com/josephroberts/edge-node-manager/device"
+)
+
+/*
+Uses the tiedot database
+https://github.com/HouzuoGuo/tiedot
+*/
+
+var directory string
+var connection *db.DB
+
+func LoadDevices(uuid string, t *device.Type) (map[int]device.Interface, error) {
+	initialise()
+	query := fmt.Sprintf(`[{"eq": "%s", "in": ["ApplicationUUID"]}]`, uuid)
+	var q interface{}
+	if err := json.Unmarshal([]byte(query), &q); err != nil {
+		return nil, err
+	}
+
+	queryResult := make(map[int]struct{})
+	devices := connection.Use("Devices")
+	if err := db.EvalQuery(q, devices, &queryResult); err != nil {
+		return nil, err
+	}
+
+	result := make(map[int]device.Interface)
+	for key := range queryResult {
+		if device, err := loadDevice(key, t); err != nil {
+			return nil, err
+		} else {
+			result[key] = device
+		}
+	}
+
+	return result, nil
+}
+
+func SaveDevice(d device.Interface) (int, error) {
+	devices := connection.Use("Devices")
+	return devices.Insert(interfaceToMap(d))
+}
+
+func RemoveDevice(key int) error {
+	devices := connection.Use("Devices")
+	return devices.Delete(key)
+}
+
+func UpdateDevices(d map[int]device.Interface) error {
+	initialise()
+	for key, value := range d {
+		if err := updateDevice(key, value); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func Stop() {
+	if connection != nil {
+		if err := connection.Close(); err != nil {
+			log.Fatalf("Unable to close database connection: %v", err)
+		}
+	}
+}
+
+func initialise() {
+	if directory == "" {
+		directory = config.GetDbDirectory()
+	}
+	if connection == nil {
+		var err error
+		if connection, err = db.OpenDB(directory); err != nil {
+			log.Fatalf("Unable to open database connection: %v", err)
+		} else {
+			collection := createCollection("Devices")
+			createIndex("ApplicationUUID", collection)
+		}
+		log.Printf("Opened database connection to %s", directory)
+	}
+}
+
+func createCollection(name string) *db.Col {
+	exists := false
+	collections := connection.AllCols()
+	for _, collection := range collections {
```

Comment:
Do we really have to do a linear search here?? Does tiedot not allow for queries? A 'database' that does let you search for things is not much of a database. Again this is going to be slow + is just weird. Surely something like [Lookup()](https://godoc.org/github.com/HouzuoGuo/tiedot/db#Lookup) could be used? If it genuinely doesn't let you search for things, even as a hash, maybe a change in DB or some cache code should be considered?

This is a repeated pattern, I definitely recommend playing with hash tables and examining their use for lookup in a loop.